main.cpp

```diff
@@ -108,7 +108,410 @@ write 3 UDTs below that EACH have:
+    void printClassDetails();
+};
+//--
+int Classroom::Student::studentId = 0;
+//--
+Classroom::Student::Student(std::string n, int a) : name(n), age(a)
+{
+    studentId++;
+}
+//--
+Classroom::Student::~Student()
+{
+    std::cout << "Student destructor called!" << std::endl;
+}
+//--
+void Classroom::Student::addGrade(double grade) 
+{
+    grades.push_back(grade);
+}
+//--
+void Classroom::Student::printDetails() const
+{
+    std::cout << "Student ID: " << studentId << ", Name: " << name << ", Age: " << age << std::endl;
+    std::cout << "Grades: ";
+    for (double grade : grades) 
+    {
+        std::cout << grade << " ";
+    }
+    std::cout << "\nAverage Grade: " << calculateAverageGrade() << std::endl;
+}
+//--
+void Classroom::Student::setFinalGrade()
+{
+    finalGrade = convertGradeIntoLetterGrade(Student::calculateAverageGrade());
+}
+//--
+double Classroom::Student::calculateAverageGrade() const{
+    if (grades.empty())
+    {
+        return 0.0;
+    }
+    double sum = 0.0;
+    for (double grade : grades) 
+    {
+        sum += grade;
+    }
+    return sum / static_cast<double>(grades.size());
+}
+//--
+Classroom::Classroom(std::string name) : 
+    classroomName(name)
+{
+}
+//--
+Classroom::~Classroom()
+{
+    std::cout << "Classroom destructor called!" << std::endl;
+}
+//--
+void Classroom::registerStudent(const Student& student) 
+{
+    students.push_back(student);
+    totalStudents++;
+}
+//--
+void Classroom::calculateAverageGrade()
+{
+    if (totalStudents == 0)
+    {
+        averageGrade = 0;
+        averageLetterGrade = 'F';
+        return;
+    }
+    double sum = 0.0;
+    for (const auto& student : students)
+    {
+        sum += student.calculateAverageGrade();
+    }
+    averageGrade = sum / totalStudents;
+    averageLetterGrade = convertGradeIntoLetterGrade(averageGrade);
+}
+//--
+void Classroom::printClassDetails(){
+    std::cout << "Classroom: " << classroomName << ", Total Students: " << totalStudents << std::endl;
+    std::cout << "Class Average Grade: " << averageGrade << " (" << averageLetterGrade << ")" << std::endl;
+    for (const Student& student : students) 
+    {
+        student.printDetails();
+    }
+}
+//--
+char Classroom::convertGradeIntoLetterGrade(double grade)
+{
+    if (grade >= 90) return 'A';
+    else if (grade >= 80) return 'B';
+    else if (grade >= 70) return 'C';
+    else if (grade >= 60) return 'D';
+    else return 'F';
+}
+//--
+struct RecreationCenter 
+{
+    Classroom academicClassroom;
+    Gym fitnessGym;
+
+    RecreationCenter(const Classroom& classroom, const Gym& gym);
+    ~RecreationCenter();
+
+    double compareAverageAge() const; //returns difference between average age of students in academic classroom and average age of members in fitness gym
+    void listUniqueNames() const;
+
+};
+//--
+RecreationCenter::RecreationCenter(const Classroom& classroom, const Gym& gym)
+    : academicClassroom(classroom), fitnessGym(gym)
```

Comment:
the const keyword and references have not been introduced yet.
please refrain from using language features that haven't been introduced in the course yet. 
- [ ] revise